core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombineSequence.java

```diff
@@ -0,0 +1,315 @@
+        + "Each node of the combining tree requires a single thread. "
+        + "Try increasing druid.processing.numThreads or "
+        + "reducing numBrokerParallelCombineThreads[%d] in the query context for a smaller tree",
+        numAvailableThreads
+    );
+  }
+
+  private static int computeNumRequiredThreads(int numChildNodes, int combineDegree)
+  {
+    // numChildNodes is used to determine that the last node is needed for the current level.
+    // Please see ParallelCombines.buildCombineTree() for more details.
+    final int numChildrenForLastNode = numChildNodes % combineDegree;
+    final int numCurLevelNodes = numChildNodes / combineDegree + (numChildrenForLastNode > 1 ? 1 : 0);
+    final int numChildOfParentNodes = numCurLevelNodes + (numChildrenForLastNode == 1 ? 1 : 0);
+
+    if (numChildOfParentNodes == 1) {
+      return numCurLevelNodes;
+    } else {
+      return numCurLevelNodes +
+             computeNumRequiredThreads(numChildOfParentNodes, combineDegree);
+    }
+  }
+
+  private Pair<Sequence<T>, Future> runCombine(
+      ReferenceCountingResourceHolder processingThreadHolder,
+      List<Sequence<T>> sequenceList
+  )
+  {
+    final Sequence<? extends Sequence<T>> sequences = Sequences.simple(sequenceList);
+    final CombiningSequence<T> combiningSequence = CombiningSequence.create(
+        new MergeSequence<>(ordering, sequences),
+        ordering,
+        combineFn
+    );
+
+    final BlockingQueue<ValueHolder> queue = new ArrayBlockingQueue<>(queueSize);
+
+    final Future future = exec.submit(
+        new PrioritizedRunnable()
+        {
+          @Override
+          public int getPriority()
+          {
+            return queryPriority;
+          }
+
+          @Override
+          public void run()
+          {
+            try {
+              combiningSequence.accumulate(
+                  queue,
+                  (theQueue, v) -> {
+                    try {
+                      addToQueue(theQueue, new ValueHolder(v));
+                    }
+                    catch (InterruptedException e) {
+                      throw new RuntimeException(e);
+                    }
+                    return theQueue;
+                  }
+              );
+              // add a null to indicate this is the last one
+              addToQueue(queue, new ValueHolder(null));
+            }
+            catch (InterruptedException e) {
+              throw new RuntimeException(e);
+            }
+            finally {
+              processingThreadHolder.close();
+            }
+          }
+        }
+    );
+
+    final Sequence<T> backgroundCombineSequence = new BaseSequence<>(
+        new IteratorMaker<T, Iterator<T>>()
+        {
+          @Override
+          public Iterator<T> make()
+          {
+            return new Iterator<T>()
+            {
+              private T nextVal;
+
+              @Override
+              public boolean hasNext()
```

Comment:
this should be true as long as `next()` is called up to `sequenceList.size()` times right? do you need to block during the `hasNext` call vs in the `next` call?

Reply:
No, all values from sequences of `sequenceList` are combined and stored in `queue`. `hasNext()` should return true until all values are aggregated and this queue becomes empty.
We can't block in `next()` since we can't know how many values will be in `queue`.

Reply:
I tried looking through the code base, and other code bases, for some direction on if blocking on `hasNext` is good or bad. I didn't find strong evidence either way. Luckily, iterators commonly have a loop like

```java
while (it.hasNext())
{
  Object next = it.next();
}
```

But from a implementation compliant standpoint, the following code yield the same result:


```java
while (it.hasNext() && it.hasNext() && it.hasNext() && it.hasNext() && it.hasNext())
{
  Object next = it.next();
}
```

Which I think is violated by this implementation